databaseV1.php

```diff
@@ -0,0 +1,375 @@
+<?php
+
+class Database extends PDO
+{
+
+	private $opt = [
+	    PDO::ATTR_ERRMODE            => PDO::ERRMODE_EXCEPTION,
+	    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
+	    PDO::ATTR_EMULATE_PREPARES   => false //sql injection flag
+	];
+
+	private $action = '';
+	private $params = [];
+  	private $cachedQueries = [];
+
+	public function __construct($driver, $charset, $host, $db, $user, $pass) {
+		parent::__construct("$driver:host=$host;dbname=$db;charset=$charset", $user, $pass, $this->opt);
+		return $this;
+	}
+
+	/**
+	 * Insert one or more rows
+	 *
+	 * @param string   	the table in which rows will be inserted
+	 * @param array 	the data to insert
+	 */ 
+	public function insert($table = '', $params = []) {
+		
+		if (!empty($params)) {
+			$this->params = $params;
+		}
+
+		// if no table or data to insert is given don't proceed
+		if ($this->missingArguments($table) || $this->missingArguments($params)) {
+			echo 'Missing insert arguments' . PHP_EOL;
+			die;
+		}
+
+		try {
+			//open transaction
+			$this->beginTransaction();
+
+			//build the query, bind values with placeholders and execute
+			foreach ($this->params as $pair) {
+				if (is_array($pair)) {
+					$theValues = str_repeat('?,', count(array_values($pair)) - 1) . '?';
+					$this->queryString = "INSERT INTO {$table} (" . implode(', ', array_keys($pair)) . ") VALUES ({$theValues})";
+		
+					$prepared = $this->prepare($this->queryString);
+					$prepared->execute(array_values($pair));
+				}
+			}
+
+			//commit transaction
+			$this->commit();
+
+			echo "Rows inserted successfully" . PHP_EOL;
+
+		} catch (PDOException $e) {
+			//rollback changes if something goes wrong
+			$this->rollback();
+			echo 'Insert failed: ' . $e->getMessage() . PHP_EOL;
```

Comment:
@tavantzo if I am not mistaken pdo already supports transactions, so I thought that transactions should be implemented like that for the task

Reply:
@MOrfanos Well, this is not a proper way for using database transactions. Can you tell why?

Reply:
@MOrfanos What will happen if do the following
```#!php
<?php

$dbm = new Database('mysql', 'utf8mb4', '127.0.0.1', 'pph', 'root', '');
$dbm->beginTransction();
$dbm->insert(....)
$dbm->insert(....)
$dbm->update(....)
$dbm->delete(....)
$dbm->select(....)
.....
```

Reply:
@tavantzo Ok I see. About the example you wrote, since it's a transaction with several statements everything will execute on commit(), so I'm thinking that inserting something and using it in the same transaction will result in an error

Reply:
@MOrfanos Well the correct answer here is that since I'm using a MySQL connection and MySQL doesn't support nested transactions an Exception will be thrown on the very first insert method call because internally it tries to start another transaction.
FYI:: Usually,  records added or modified within a transaction are fully accessible by the connection that had started it, while they are "invisible" for other connections until the transaction commits.

Reply:
@tavantzo ok i see, thank you